test/unit/auth/auth.spec.ts

```diff
@@ -1401,10 +1407,204 @@ AUTH_CONFIGS.forEach((testConfig) => {
+            deleteProvider: [ 'phone', 'google.com' ],
           });
       });
 
+      describe('non-federated providers', () => {
+        let invokeRequestHandlerStub: sinon.SinonStub;
+        let getAccountInfoByUidStub: sinon.SinonStub;
+        beforeEach(() => {
+          invokeRequestHandlerStub = sinon.stub(testConfig.RequestHandler.prototype, 'invokeRequestHandler')
+            .resolves({
+              // nothing here is checked; we just need enough to not crash.
+              users: [{
+                localId: 1,
+              }],
+            });
+
+          getAccountInfoByUidStub = sinon.stub(testConfig.RequestHandler.prototype, 'getAccountInfoByUid')
+            .resolves({
+              // nothing here is checked; we just need enough to not crash.
+              users: [{
+                localId: 1,
+              }],
+            });
+        });
+        afterEach(() => {
+          invokeRequestHandlerStub.restore();
+          getAccountInfoByUidStub.restore();
+        });
+
+
+        it('specifying both email and providerId=email should be rejected', () => {
+          expect(() => {
+            auth.updateUser(uid, {
+              email: 'user@example.com',
+              providerToLink: {
+                providerId: 'email',
+                uid: 'user@example.com',
+              },
+            });
+          }).to.throw(FirebaseAuthError).with.property('code', 'auth/argument-error');
+        });
+
+        it('specifying both phoneNumber and providerId=phone should be rejected', () => {
+          expect(() => {
+            auth.updateUser(uid, {
+              phoneNumber: '+15555550001',
+              providerToLink: {
+                providerId: 'phone',
+                uid: '+15555550001',
+              },
+            });
+          }).to.throw(FirebaseAuthError).with.property('code', 'auth/argument-error');
+        });
+
+        it('email linking should use email field', async () => {
+          await auth.updateUser(uid, {
+            providerToLink: {
+              providerId: 'email',
+              uid: 'user@example.com',
+            },
+          });
+          expect(invokeRequestHandlerStub).to.have.been.calledOnce.and.calledWith(
+            sinon.match.any, sinon.match.any, {
+              localId: uid,
+              email: 'user@example.com',
+            });
+        });
+
+        it('phone linking should use phoneNumber field', async () => {
+          await auth.updateUser(uid, {
+            providerToLink: {
+              providerId: 'phone',
+              uid: '+15555550001',
+            },
+          });
+          expect(invokeRequestHandlerStub).to.have.been.calledOnce.and.calledWith(
+            sinon.match.any, sinon.match.any, {
+              localId: uid,
+              phoneNumber: '+15555550001',
+            });
+        });
+
+        it('specifying both phoneNumber=null and providersToDelete=phone should be rejected', () => {
```

Comment:
nit: does this need to be reject?

Reply:
I don't know. It does indicate a slight misuse of our API as there's not really any situation where the user should be deleting both by setting phoneNumber=null and also by adding phone to providersToDelete. OTOH, the user's intent is clear in this case. See also the comment in the implementation (src/auth/auth.ts:updateUser:~313)

I'm inclined to be a bit conservative and reject this situation as an error and then relax it later if necessary. But am also willing to be convinced otherwise. We'll get a few more reviews before this lands, so we can probably just go with the consensus that results from that.

I've left this alone for now.